students/timm67/session05/mailroom_part4.py

```diff
@@ -0,0 +1,192 @@
+        donor = (donor_fname, donor_lname)
+        donors[donor] = donor_amt
+
+
+def generate_thankyou_single_donor_task(donor):
+    global donors
+    total_donations = float(0.0)
+    outlines = []
+    fmtline0 = "\n\nDear {0},\n\n"
+    fmtline1 = "Many thanks for your recent donation of {0:6.2f}.\n\n"
+    fmtline2 = "Your total donations are {0:6.2f}, averaging {1:6.2f}\n\n"
+    fmtline3 = "Thanks for your generous donation.\n\nBest Regards,\nThe Staff"
+    donations = donors[donor]
+    for donation in donations:
+        total_donations += donation
+    avg_donation = total_donations / len(donations)
+    outlines.append(fmtline0.format(donor[0]))
+    outlines.append(fmtline1.format(donations[(len(donations) - 1)]))
+    outlines.append(fmtline2.format(total_donations, avg_donation))
+    outlines.append(fmtline3.format())
+    [print(outline) for outline in outlines]
+    return outlines
+
+
+def send_thankyou_multiple_donors_task():
+    global donors
+    filenames = []
+    ffmtline0 = "\n\nDear {0},\n\n"
+    ffmtline1 = "Many thanks for your recent donation of {0:6.2f}.\n\n"
+    ffmtline2 = "Your total donations are {0:6.2f}, averaging {1:6.2f}\n\n"
+    ffmtline3 = "Thanks for your generous donation.\nBest Regards,\nThe Staff"
+    for donor in donors.keys():
+        filename = "./{0}_{1}.txt".format(donor[0], donor[1])
+        filenames.append(filename)
+        print("Writing {0}...".format(filename))
+        donations = donors[donor]
+        total_donations = sum(donations)
+        avg_donation = total_donations / len(donations)
+        try:
+            with open(filename, 'w') as handle:
+                handle.write(ffmtline0.format(donor[0]))
+                handle.write(ffmtline1.format(donations[(len(donations) - 1)]))
+                handle.write(ffmtline2.format(total_donations, avg_donation))
+                handle.write(ffmtline3.format())
+        except PermissionError:
+            print("Permission Error")
+        except IOError:
+            print("I/O Error")
+    return filenames
+
+
+def formatter(fname, lname, in_list):
+    list_len = len(in_list)
+    fmt_string = "{0:15} {1:15} ".format(fname, lname)
+    fmt_string += list_len * '{:6.2f}, '
+    return str(fmt_string.format(*in_list))
+
+
+def create_report_task():
+    global donors
+    outlines = []
+    report_fmt_str_hdr = "{0:15} {1:15} {2:10}"
+    print(report_fmt_str_hdr.format('First Name', 'Last Name', ' Donations'))
+    for key in donors.keys():
+        outlines.append(formatter(key[0], key[1], donors[key]))
+    [print(outline) for outline in outlines]
+    return outlines
+
+
+def print_menu():
+    print("Mailroom Tasks")
+    print("[1] Send a Thank you to a single donor")
+    print("[2] Create a report")
+    print("[3] Send letters to all donors")
+    print("[9] Exit Mailroom")
+    pass
+
+
+def exit_mail():
+    global exitMail
+    exitMail = True
```

Comment:
Minor: Not very desirable to rely on a global variable. Could return some value and let the main loop receive/detect the special value for exit. Also could just `exit()` here.